File: part4/exercise.py
"""
    1. Write a grammar that describes arithmetic expressions containing any number of +, -, *, or / operators.
        expr    :   term   ( ( ADD | SUB ) term   )*
        term    :   factor ( ( MUL | DIV ) factor )*
        factor  :   INTEGER
    
    2. Using the grammar, write an interpreter that can evaluate arithmetic expressions containing any number 
       of +, -, *, or / operators. Your interpreter should be able to handle expressions like 
       “2 + 7 * 4”, “7 - 8 / 4”, “14 + 2 * 3 - 6 / 2”, and so on.

"""

# This intepreter takes any expression with the format: INTEGER ((ADD, SUB, DIV, MUL) INTEGER)*

# Tokentypes
EOF, ADD, SUB, MUL, DIV, INTEGER = 'EOF', 'ADD', 'SUB', 'MUL', 'DIV', 'INTEGER'

# Class Token
class Token(object):
    def __init__(self, type, value):
        self.type = type
        self.value = value

# Class Lexer
class Lexer(object):
    def __init__(self, text):
        self.text = text
        self.pos = 0 # current input character pointer
        self.current_char = self.text[self.pos]
    
    def error():
        raise Exception('Invalid character')

    def advance(self):
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()
    
    def integer(self):
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        
        return int(result)

    # get the next token on demand.
    def get_next_token(self):
        if self.current_char is None:
            return Token(EOF, None)

        if self.current_char.isspace():
            self.skip_whitespace()
        
        if self.current_char.isdigit():
            return Token(INTEGER, self.integer())
        
        if self.current_char == '+':
            self.advance()
            return Token(ADD, '+')
        
        if self.current_char == '-':
            self.advance()
            return Token(SUB, '-')
        
        if self.current_char == '*':
            self.advance()
            return Token(MUL, '*')
        
        if self.current_char == '/':
            self.advance()
            return Token(DIV, '/')
        
        self.error()

# class Parser
class Parser(object):
    def __init__(self, lexer):
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token() # the look ahead token.
    
    def error():
        raise Exception('Invalid Syntax')

    def expr(self):
        result = self.term()
        while self.current_token.type in (ADD, SUB):
            token = self.current_token
            if token.type == ADD:
                self.eat(ADD)
                result += self.term()
            elif token.type == SUB:
                self.eat(SUB)
                result -= self.term()
        return result

    def term(self):
        result = self.factor()
        while self.current_token.type in (MUL, DIV):
            token = self.current_token
            if token.type == MUL:
                self.eat(MUL)
                result *= self.factor()
            elif token.type == DIV:
                self.eat(DIV)
                result /= self.factor()
        return result
    
    def factor(self):
        value = self.current_token.value
        self.eat(INTEGER)
        return value
    
    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error()

def main():
    
    while True:
        try:
            text = input('calc> ')
        except EOFError:
            break
        if not text:
            continue
        lexer = Lexer(text)
        parser = Parser(lexer)
        result = parser.expr()
        print(result)

    """
    text = "5*5"
    lexer = Lexer(text)
    parser = Parser(lexer)
    result = parser.expr()
    print(result)    
    """

if __name__ == '__main__':
    main()

    

File: part6/exercise.go
package main

// Write your own version of the interpreter of arithmetic expressions
// as described in this article.
// Remember: repetition is the mother of all learning.

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const (
	INTEGER = "INTEGER"
	PLUS    = "PLUS"
	MINUS   = "MINUS"
	MUL     = "MUL"
	DIV     = "DIV"
	LPAREN  = "LPAREN"
	RPAREN  = "RPAREN"
	EOF     = "EOF"
	NONE    = 255
)

type TokenType string

type Token struct {
	Type  TokenType
	Value interface{}
}

type Lexer struct {
	input       string
	currentPos  int
	currentChar byte
}

func NewLexer(input string) *Lexer {
	l := &Lexer{input: input}
	l.currentPos = 0
	l.currentChar = l.input[l.currentPos]
	return l
}

func (l *Lexer) error() {
	fmt.Printf("Unknown character %c\n", l.currentChar)
	os.Exit(1)
}

func (l *Lexer) advance() {
	l.currentPos++
	if l.currentPos > len(l.input)-1 {
		l.currentChar = NONE
	} else {
		l.currentChar = l.input[l.currentPos]
	}
}

func (l *Lexer) skipWhiteSpace() {
	for l.isSpace(l.currentChar) && l.currentChar != NONE {
		l.advance()
	}
}

func (l *Lexer) integer() int {
	var result string
	for l.isDigit(l.currentChar) && l.currentChar != NONE {
		result += string(l.currentChar)
		l.advance()
	}
	num, _ := strconv.Atoi(result)
	return num
}

func (l *Lexer) getNextToken() *Token {
	for l.currentChar != NONE {
		if l.isSpace(l.currentChar) {
			l.skipWhiteSpace()
		}
		if l.isDigit(l.currentChar) {
			return &Token{Type: INTEGER, Value: l.integer()}
		}
		if l.currentChar == '+' {
			l.advance()
			return &Token{Type: PLUS, Value: "+"}
		}
		if l.currentChar == '-' {
			l.advance()
			return &Token{Type: MINUS, Value: "-"}
		}
		if l.currentChar == '*' {
			l.advance()
			return &Token{Type: MUL, Value: "*"}
		}
		if l.currentChar == '/' {
			l.advance()
			return &Token{Type: DIV, Value: "/"}
		}
		if l.currentChar == '(' {
			l.advance()
			return &Token{Type: LPAREN, Value: "("}
		}
		if l.currentChar == ')' {
			l.advance()
			return &Token{Type: RPAREN, Value: ")"}
		}
		l.error()
	}
	return &Token{Type: EOF, Value: NONE}
}

func (l *Lexer) isSpace(ch byte) bool {
	return ch == ' '
}

func (l *Lexer) isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

// Interpreter ->
type Interpreter struct {
	lexer        *Lexer
	currentToken *Token
}

// NewInterpreter ->
func NewInterpreter(lexer *Lexer) *Interpreter {
	i := &Interpreter{lexer: lexer}
	i.currentToken = i.lexer.getNextToken()
	return i
}

func (i *Interpreter) error() {
	fmt.Printf("Syntax Error %v\n", i.currentToken)
	os.Exit(1)
}

func (i *Interpreter) eat(tokenType TokenType) {
	if i.currentToken.Type == tokenType {
		i.currentToken = i.lexer.getNextToken()
	} else {
		i.error()
	}
}

func (i *Interpreter) factor() int {
	token := i.currentToken
	var result int
	if token.Type == LPAREN {
		i.eat(LPAREN)
		result = i.expr()
		i.eat(RPAREN)
	} else {
		i.eat(INTEGER)
		result = token.Value.(int)
	}
	return result
}

func (i *Interpreter) term() int {
	result := i.factor()
	for i.currentToken.Type == MUL || i.currentToken.Type == DIV {
		if i.currentToken.Type == MUL {
			i.eat(MUL)
			result *= i.factor()
		}
		if i.currentToken.Type == DIV {
			i.eat(DIV)
			result /= i.factor()
		}
	}
	return result
}

func (i *Interpreter) expr() int {
	result := i.term()
	for i.currentToken.Type == PLUS || i.currentToken.Type == MINUS {
		if i.currentToken.Type == PLUS {
			i.eat(PLUS)
			result += i.term()
		}
		if i.currentToken.Type == MINUS {
			i.eat(MINUS)
			result -= i.term()
		}
	}
	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("calc> ")
		scanned := scanner.Scan()
		if !scanned {
			continue
		}
		input := scanner.Text()
		lexer := NewLexer(input)
		interpreter := NewInterpreter(lexer)
		result := interpreter.expr()
		fmt.Printf("%d\n", result)
	}
}
